### ISSUE TITLE
Test d3 is flaky due to Module Not Found d3

### ISSUE THREAD

**Issue by FabioRosado** [pyscript/pyscript#759]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a duplication
- [X] I already searched in Google and didn't find any good information or help

### What is the issue/comment/problem?

Seems like our integration test for `d3` is flaky, it often fails with a `ModuleNotFound` error.

```
________________________ TestExamples.test_d3[chromium] ________________________
[77](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:78)

[78](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:79)
self = <integration.test_zz_examples.TestExamples object at 0x7fc79a5d26a0>
[79](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:80)

[80](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:81)
    def test_d3(self):
[81](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:82)
        # XXX improve this test
[82](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:83)
        self.goto("examples/d3.html")
[83](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:84)
>       self.wait_for_pyscript()
[84](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:85)

[85](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:86)
tests/integration/test_zz_examples.py:111: 
[86](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:87)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[87](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:88)
tests/integration/support.py:189: in wait_for_pyscript
[88](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:89)
    self.wait_for_console(
[89](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:90)
tests/integration/support.py:168: in wait_for_console
[90](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:91)
    pass
[91](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:92)
env/lib/python3.9/site-packages/playwright/_impl/_sync_base.py:78: in __exit__
[92](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:93)
    self._event.value
[93](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:94)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[94](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:95)

[95](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:96)
self = <playwright._impl._sync_base.EventInfo object at 0x7fc79648cc10>
[96](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:97)

[97](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:98)
    @property
[98](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:99)
    def value(self) -> T:
[99](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:100)
        while not self._future.done():
[100](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:101)
            self._sync_base._dispatcher_fiber.switch()
[101](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:102)
        asyncio._set_running_loop(self._sync_base._loop)
[102](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:103)
        exception = self._future.exception()
[103](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:104)
        if exception:
[104](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:105)
>           raise exception
[105](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:106)
E           playwright._impl._api_types.TimeoutError: Timeout 30000.0ms exceeded while waiting for event "console"
[106](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:107)
E           =========================== logs ===========================
[107](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:108)
E           waiting for event "console"
[108](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:109)
E           ============================================================
[109](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:110)

[110](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:111)
env/lib/python3.9/site-packages/playwright/_impl/_sync_base.py:58: TimeoutError
[111](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:112)
---------------------------- Captured stdout setup -----------------------------
[112](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:113)

[113](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:114)
----------------------------- Captured stdout call -----------------------------
[114](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:115)
[  0.00 page.goto      ] examples/d3.html
[115](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:116)
[  0.01 http_server    ] "GET /examples/d3.html HTTP/1.1" 200 -
[116](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:117)
[  0.02 http_server    ] "GET /build/pyscript.css HTTP/1.1" 200 -
[117](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:118)
[  0.02 http_server    ] "GET /build/pyscript.js HTTP/1.1" 200 -
[118](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:119)
[  0.20 console.info   ] [pyscript/main] checking for py-confing
[119](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:120)
[  0.20 console.info   ] [py-config] config set: {autoclose_loader: true, runtimes: Array(1)}
[120](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:121)
[  0.20 console.info   ] [py-config] Initializing runtimes
[121](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:122)
[  0.20 console.info   ] [pyscript/pyodide] Runtime config: {name: pyodide-default, lang: python, src: [https://cdn.jsdelivr.net/pyodide/v0.21.2/full/pyodide.js}](https://cdn.jsdelivr.net/pyodide/v0.21.2/full/pyodide.js%7D)
[122](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:123)
[  0.20 console.info   ] [pyscript/main] add py-loader
[123](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:124)
[  0.25 console.info   ] [py-loader] Loading runtime...
[124](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:125)
[  0.25 console.info   ] [pyscript/pyodide] Loading pyodide
[125](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:126)
[  2.87 console.log    ] Python initialization complete
[126](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:127)
[  2.87 console.info   ] [pyscript/pyodide] pyodide.loadPackage: micropip
[127](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:128)
[  2.87 console.info   ] [pyscript/pyodide] Loading micropip, pyparsing, packaging, distutils
[128](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:129)
[  3.10 console.info   ] [pyscript/pyodide] Loaded pyparsing, packaging, micropip, distutils
[129](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:130)
[  3.10 console.info   ] [pyscript/pyodide] importing pyscript.py
[130](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:131)
[  4.03 console.info   ] [pyscript/pyodide] pyodide loaded and initialized
[131](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:132)
[  4.03 console.info   ] [py-loader] Runtime created...
[132](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:133)
[  4.03 console.info   ] [py-loader] Initializing components...
[133](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:134)
[  4.03 console.info   ] [py-script] py-mount: found 0 elements
[134](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:135)
[  4.03 console.info   ] [py-loader] Initializing scripts...
[135](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:136)
[  4.06 console.error  ] [pyscript/base] PythonError: Traceback (most recent call last):
[136](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:137)
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
[137](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:138)
    raise self._exception
[138](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:139)
  File "/lib/python3.10/asyncio/tasks.py", line 232, in __step
[139](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:140)
    result = coro.send(None)
[140](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:141)
  File "/lib/python3.10/site-packages/_pyodide/_base.py", line 506, in eval_code_async
[141](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:142)
    await CodeRunner(
[142](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:143)
  File "/lib/python3.10/site-packages/_pyodide/_base.py", line 357, in run_async
[143](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:144)
    coroutine = eval(self.code, globals, locals)
[144](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:145)
  File "<exec>", line 2, in <module>
[145](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:146)
ModuleNotFoundError: No module named 'd3'
[146](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:147)

[147](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:148)
[  4.07 console.info   ] [py-loader] Running post initializers...
[148](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:149)
[  4.07 console.info   ] [py-loader] Closing
[149](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:150)
[  4.08 console.debug  ] [py-script] Initializing py-* event handlers...
[150](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:151)
[  4.08 console.info   ] [pyscript/runtime] PyScript page fully initialized
[151](https://github.com/pyscript/pyscript/runs/8211017933?check_suite_focus=true#step:10:152)
=========================== short test summary info ============================
```

**Comment by FabioRosado** [comment]
Just poked at it a bit and it seems that if we fix the URL the d3 example work. 

I know I've seen this working in the past, so I'm not sure if this was just some recent change in d3/skypack itself? 
Either way replacing `https://cdn.skypack.dev/d3@7` with `https://cdn.skypack.dev/pin/d3@v7.6.1-1Q0NZ0WZnbYeSjDusJT3/mode=imports,min/optimized/d3.js` seems to have fixed the problem

**Comment by marimeireles** [comment]
Hey Fabio! It seems like this issue is caused by some upstream problem in pyodide that was fixed. We just need to update the version we're using fir Pyscript and this should be good again.

**Comment by marimeireles** [comment]
But it's puzzling that your solution also works 
Maybe we wait for the Pyodide thing just to be sure?

**Comment by FabioRosado** [comment]
Hey Mariana!  That's good to know! I'm assuming that by providing the URL that contains the js file, pyodide is probably just importing that file and that's why it works :thinking: 

I'll revert the changes in my PR to add tests since I don't think it's worth blocking it for this one test :smile:

**Comment by marimeireles** [comment]
I think this can be closed as complete as your PR was merged?

**Comment by antocuni** [comment]
I'm reopening this issue because it seems that the test is flaky again.
I will commit the relevant `@pytest.mark.skip` as part of (unrelated) PR #874 .

> Either way replacing `https://cdn.skypack.dev/d3@7` with `https://cdn.skypack.dev/pin/d3@v7.6.1-1Q0NZ0WZnbYeSjDusJT3/mode=imports,min/optimized/d3.js` seems to have fixed the problem

I don't really know, but this is my wild guess at what's happening:
1. `register_js_importmap` (formerly known as `register_esm`) looks for importmaps in the the DOM, then calls `runtime.registerJSModule`
2. then, pyodide does some magic to make the registered modules available as `import xxx`
3. if we run `import xxx` before pyodide has done its magic and/or the browser has completed the download of the module, we get an `ImportError`.

So, using a minified version of `d3` probably made step (2) faster, thus mitigating the issue but without really solving it. And my recent changes in the life cycle probably caused the rest of the startup to be faster so that now you get the error again :).

But as I said this is just a wild guess, we should investigate more.

**Comment by chrisamow** [comment]
> I'm reopening this issue because it seems that the test is flaky again. I will commit the relevant `@pytest.mark.skip` as part of (unrelated) PR #874 .
> 
> > Either way replacing `https://cdn.skypack.dev/d3@7` with `https://cdn.skypack.dev/pin/d3@v7.6.1-1Q0NZ0WZnbYeSjDusJT3/mode=imports,min/optimized/d3.js` seems to have fixed the problem
> 
> I don't really know, but this is my wild guess at what's happening:
> 
>     1. `register_js_importmap` (formerly known as `register_esm`) looks for importmaps in the the DOM, then calls `runtime.registerJSModule`
> 
>     2. then, pyodide does some magic to make the registered modules available as `import xxx`
> 
>     3. if we run `import xxx` before pyodide has done its magic and/or the browser has completed the download of the module, we get an `ImportError`.
> 
> 
> So, using a minified version of `d3` probably made step (2) faster, thus mitigating the issue but without really solving it. And my recent changes in the life cycle probably caused the rest of the startup to be faster so that now you get the error again :).
> 
> But as I said this is just a wild guess, we should investigate more.

The d3 example demo is still failing, just fyi
PythonError: Traceback (most recent call last): File "/lib/python3.10/asyncio/futures.py", line 201, in result raise self._exception File "/lib/python3.10/asyncio/tasks.py", line 232, in __step result = coro.send(None) File "/lib/python3.10/site-packages/_pyodide/_base.py", line 506, in eval_code_async await CodeRunner( File "/lib/python3.10/site-packages/_pyodide/_base.py", line 357, in run_async coroutine = eval(self.code, globals, locals) File "", line 2, in ModuleNotFoundError: No module named 'd3'

![image](https://user-images.githubusercontent.com/5819969/205175666-309b0140-0459-448b-97a5-5cc5b24dbef0.png)

pyscript is pretty sweet - i've been looking to escape js for so long :)

**Comment by antocuni** [comment]
@chrisamow thank you for reopening the issue, it was closed by mistake (I guess it was closed automatically by github but I didn't really want to).
The core of the problem is that our `importmap` support is completely broken, as explained here:
https://github.com/pyscript/pyscript/blob/5aa9135a344822bf2e250a0bfbc78f58bdab74bb/pyscriptjs/src/plugins/importmap.ts#L14-L23

**Comment by tedpatrick** [comment]
Fixed in #1025
